File: react-web/library/src/components/TspTree.js
import { React, h, cx, themed, starPlanetTheme } from './_shared.js';
import { useState } from 'react';

function nodeId(node, index) {
  return String(node?.id ?? node?.key ?? index);
}

function nodeLabel(node) {
  return node?.label ?? node?.title ?? node?.text ?? String(node?.id ?? '');
}

function TreeNode({ node, depth, selectedId, expanded, toggle, onSelect, index }) {
  const id = nodeId(node, index);
  const kids = Array.isArray(node?.children) ? node.children : [];
  const open = expanded.has(id);
  return h(
    'div',
    { className: 'bc-tree__node', key: id },
    h(
      'div',
      {
        className: cx('bc-tree__row', selectedId === id && 'bc-selected'),
        style: { paddingLeft: 8 + depth * 16 },
      },
      kids.length
        ? h(
            'button',
            {
              type: 'button',
              className: cx('bc-tree__twist', open && 'bc-open'),
              'aria-label': open ? 'Collapse' : 'Expand',
              onClick: () => toggle(id),
            },
            open ? '▾' : '▸'
          )
        : h('span', { className: 'bc-tree__twist bc-tree__twist--leaf' }),
      h(
        'button',
        {
          type: 'button',
          className: 'bc-tree__label',
          onClick: () => onSelect?.(id, node),
        },
        nodeLabel(node)
      )
    ),
    open
      ? kids.map((child, i) =>
          h(TreeNode, {
            key: nodeId(child, i),
            node: child,
            depth: depth + 1,
            selectedId,
            expanded,
            toggle,
            onSelect,
            index: i,
          })
        )
      : null
  );
}

/** Nested tree selector. */
export function TspTree({
  items = [],
  selectedId,
  expandedIds,
  theme = starPlanetTheme,
  onSelect,
  onExpand,
}) {
  const list = Array.isArray(items) ? items : [];
  const [localExpanded, setLocalExpanded] = useState(() => new Set(expandedIds || []));
  const expanded = expandedIds != null ? new Set(expandedIds) : localExpanded;

  const toggle = (id) => {
    const next = new Set(expanded);
    if (next.has(id)) next.delete(id);
    else next.add(id);
    if (expandedIds == null) setLocalExpanded(next);
    onExpand?.(Array.from(next));
  };

  return h(
    'div',
    { className: 'bc-tree', style: themed(theme), role: 'tree' },
    list.map((node, i) =>
      h(TreeNode, {
        key: nodeId(node, i),
        node,
        depth: 0,
        selectedId,
        expanded,
        toggle,
        onSelect,
        index: i,
      })
    )
  );
}

File: ios-swiftui/library/Sources/PlanetComponents/TspTree.swift
import SwiftUI

public struct TspTreeNode: Identifiable, Equatable {
    public var id: String
    public var label: String
    public var children: [TspTreeNode]

    public init(id: String, label: String, children: [TspTreeNode] = []) {
        self.id = id
        self.label = label
        self.children = children
    }
}

/// Nested tree selector.
public struct TspTree: View {
    public var items: [TspTreeNode]
    public var selectedId: String?
    public var expandedIds: [String]?
    public var theme: StarPlanetTheme
    public var onSelect: (String, TspTreeNode) -> Void
    public var onExpand: ([String]) -> Void

    @State private var localExpanded: Set<String>

    public init(
        items: [TspTreeNode] = [],
        selectedId: String? = nil,
        expandedIds: [String]? = nil,
        theme: StarPlanetTheme = .sky,
        onSelect: @escaping (String, TspTreeNode) -> Void = { _, _ in },
        onExpand: @escaping ([String]) -> Void = { _ in }
    ) {
        self.items = items
        self.selectedId = selectedId
        self.expandedIds = expandedIds
        self.theme = theme
        self.onSelect = onSelect
        self.onExpand = onExpand
        _localExpanded = State(initialValue: Set(expandedIds ?? []))
    }

    private var effectiveExpanded: Set<String> {
        if let expandedIds { return Set(expandedIds) }
        return localExpanded
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(items) { node in
                TspTreeNodeRow(
                    node: node,
                    depth: 0,
                    selectedId: selectedId,
                    expanded: effectiveExpanded,
                    theme: theme,
                    onSelect: onSelect,
                    onToggle: toggle
                )
            }
        }
        .padding(.vertical, 6)
        .background(theme.surfaceRaised)
        .overlay(RoundedRectangle(cornerRadius: 16).stroke(theme.borderDefault, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }

    private func toggle(_ id: String) {
        var next = effectiveExpanded
        if next.contains(id) {
            next.remove(id)
        } else {
            next.insert(id)
        }
        if expandedIds == nil {
            localExpanded = next
        }
        onExpand(Array(next))
    }
}

private struct TspTreeNodeRow: View {
    let node: TspTreeNode
    let depth: Int
    let selectedId: String?
    let expanded: Set<String>
    let theme: StarPlanetTheme
    let onSelect: (String, TspTreeNode) -> Void
    let onToggle: (String) -> Void

    var body: some View {
        let open = expanded.contains(node.id)
        let selected = selectedId == node.id
        let hasKids = !node.children.isEmpty

        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 0) {
                if hasKids {
                    Button(action: { onToggle(node.id) }) {
                        Text(open ? "▾" : "▸")
                            .font(.system(size: 14, weight: .heavy))
                            .foregroundColor(theme.textSecondary)
                            .frame(width: 24, alignment: .center)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(open ? "Collapse" : "Expand")
                } else {
                    Color.clear.frame(width: 24)
                }
                Button(action: { onSelect(node.id, node) }) {
                    Text(node.label)
                        .font(.system(size: 14, weight: selected ? .black : .bold))
                        .foregroundColor(theme.textPrimary)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                .buttonStyle(.plain)
            }
            .padding(.leading, 8 + CGFloat(depth) * 16)
            .padding(.trailing, 8)
            .padding(.vertical, 8)
            .background(selected ? theme.selectedFill : Color.clear)

            if open {
                ForEach(node.children) { child in
                    TspTreeNodeRow(
                        node: child,
                        depth: depth + 1,
                        selectedId: selectedId,
                        expanded: expanded,
                        theme: theme,
                        onSelect: onSelect,
                        onToggle: onToggle
                    )
                }
            }
        }
    }
}
